### PR TITLE
add signature at footer

### LINKED ISSUE #385: Allow specifying footer
It would be great, if one could specify a raw footer like you can specify a header for the print.

### PR DESCRIPTION
**Pull request by cqhaibin**
add signature at footer, and only json'data.
Signed-off-by: sam Dragon <kxhaibin@163.com>

### CODE REVIEW

**Review by AshifMohammad — APPROVED** [review]
LTGM, however - there are few conflicts , which need to be resolved.

**Comment by crabbly** [comment]
Hey @cqhaibin, I like the idea of the footer. I'm wondering if we should add for images and html as well, since the header is used on those print jobs.

If you could please fix the conflicts, I think we can merge this.

Thank you for helping with the lib.

File: test/manual/index.html [hunk 1]
```diff
@@ -75,6 +75,33 @@
         })
     }
 
+    function printJsonOffooter() {
```
Comment (crabbly) [inline test/manual/index.html:78]: If you can, please rename this to `printJsonWithFooter`

File: src/js/json.js [hunk 2]
```diff
@@ -39,6 +39,9 @@ export default {
     // Build the printable html data
     htmlData += jsonToHTML(params)
 
+    //Check if there is a footer on end of the table
+    if (params.footer) htmlData += '<div style="' + params.footerStyle + '">' + params.footer + '</div>'
```
Comment (crabbly) [inline src/js/json.js:43]: We need to standardize our conditionals style in the code base. For now, if you could please follow the style used in the header, that would be awesome.

**Comment by crabbly** [comment]
I just realized we have another PR for the same feature, with the addition of HTML support.
https://github.com/crabbly/Print.js/pull/393
I haven't had time to work on this lib in a while, that MR should have been merged already. Sorry for the delay guys.

File: test/manual/index.html (hunk 1 above)
Comment (cqhaibin) [reply test/manual/index.html:78]: ok, i agree.

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,5 @@
 .idea/
 node_modules/
-dist/
```
Comment (AshifMohammad) [inline .gitignore:3]: any reason why have we removed the `dist/` from `.gitignore`? probably you dont want to checkin the build folder ?

File: src/js/functions.js [hunk 4]
```diff
@@ -103,12 +103,48 @@ export function addHeader (printElement, header, headerStyle) {
...
+
+      params.onPrintDialogClose()
+    }
+
+    window.addEventListener(event, handler)
+  }
+}
+
+export function isRawHTML (raw) {
+  let regexHtml = new RegExp('<([A-Za-z][A-Za-z0-9]*)\\b[^>]*>(.*?)</\\1>')
```
Comment (AshifMohammad) [inline src/js/functions.js]: if `regexHtml` is not being reused can we make it `const`?

Comment (AshifMohammad) [reply src/js/functions.js]:
```javascript
export  function isRawHTML (raw) {
  const regexHtml = new RegExp('<([A-Za-z][A-Za-z0-9]*)\\b[^>]*>(.*?)</\\1>')

  return regexHtml.test(raw)
}
```
`console.log(isRawHTML(`<div> `dummy div`</div>`) => true`

File: src/js/html.js [hunk 5]
```diff
@@ -48,6 +48,11 @@ export default {
       addHeader(printableElement, params.header, params.headerStyle)
     }
 
+    // Add footer
+    if (params.footer) {
```
Comment (AshifMohammad) [inline src/js/html.js:52]: nitpick - you could have destructured it 
```javascript
 if ({footer,footerStyle }) {
       addFooter(printableElement, footer, footerStyle)
    }

File: src/js/image.js [hunk 6]
```diff
@@ -18,6 +18,11 @@ export default {
       // Check if we are adding a header
       if (params.header) addHeader(printableElement, params.header, params.headerStyle)
 
+      // Add footer
+      if (params.footer) {
```
Comment (AshifMohammad) [inline src/js/image.js:22]: nitpick - you could have destructured it
```javascript
 if ({footer,footerStyle }) {
       addFooter(printableElement, footer, footerStyle)
    }

File: test/manual/index.html [hunk 7]
```diff
@@ -96,7 +104,7 @@
         })
     }
 
-    function printJsonOffooter() {
+    function printJsonWithFooter() {
       let data = [];
       for (let i=0; i <= 1000; i++){
         data.push({
```
Comment (AshifMohammad) [inline test/manual/index.html:110]: nit pick - you can avoid mutation here and maybe use the spread to achieve the same and infact limit the loop till the length of the array?
```javascript
function printJsonWithFooter() {
  let data = [];
  for (let i=0; i <= data.length; i++){
    return [...data,{test1: createRandomString(),test2: createRandomString()}]
      
  }
}

**Review by AshifMohammad — COMMENTED** [review]
Please address the comments

File: .gitignore (hunk 3 above)
Comment (crabbly) [reply .gitignore:3]: Ashif is correct. We don't want to push `dist` into the repository.

File: src/js/html.js (hunk 5 above)
Comment (crabbly) [reply src/js/html.js:52]: Although I love JavasScript destructuring, in my opinion, using it in this conditional is unnecessary. And probably less readable too. All we wan't to validate is the `footer` property, not `footerStyle`.

**Comment by epkbr** [comment]
footer not working in firefox

**Pull request closed without merging**